### PR TITLE
Swagger documentation gsoc2019

### PR DESCRIPTION
**Pull request by kangbreder**
## Description
For GSoC, I have improved the Swagger document with the following

- Reflected all API descriptions on swagger UI

- Formatted API descriptions

- Added swagger docs for new APIs

- Added swagger-gradle-plugin for automatic save of json during build.

- Leveraged swagger codegen for generation of language-specific libraries using gradle-swagger-generator plugin.

Link to JIRA TICKET: https://jira.apache.org/jira/browse/FINERACT-733

- [COMPLETED] Complete Swagger documentation on all API’s of Fineract

- [COMPLETED] Description inside of swagger docs

- [IN PROGRESS] Try it out button for API’s from swagger-UI

- [IN PROGRESS] Automatically generate the swagger-docs -> building with Fineract

- [NOT COMPLETED] Navigation Bar on swagger-UI(For easy access of documentation of API's)

- [NOT COMPLETED] Migrating the complete overview section in swagger-docs/swagger-docs

- [COMPLETED] Swagger guide for new Api’s or how to add a swagger doc to any Fineract API

- [IN PROGRESS] QA round to check if all the API’s are reflecting or not(Correct)

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [ ] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by Sanyam96** [comment]
Adding Authentication and common params for swagger documentation to each and every endpoint(API's ) is unnecessary and is not a good coding standard.
It will increase redundant code for auth to every API's, which can be done by sharing a common resource to all the API's

In the review of removing Auth params from every API in your commit: 
https://github.com/kangbreder/fineract/commit/7418ac403e4e145c86dd826086a2799dd5c4bf7c#commitcomment-34794554

Here are some reference links which you can refer to remove those params!
https://stackoverflow.com/questions/47627469/swagger2-too-many-the-same-annotations
https://stackoverflow.com/questions/39853217/swagger-how-to-combine-shared-information-from-different-apis

**Comment by kangbreder** [comment]
Okay. I will consider that and let you know.

**Comment by Sanyam96** [comment]
I don't think, we should require the below files in the codebase
catalina.base_IS_UNDEFINED/logs/fineract-platform.log
.response.json.swp

if we need anyhow,
please try to add some documentation regarding those files

**Comment by Sanyam96** [comment]
I don't think, we should require the below files in the codebase
```catalina.base_IS_UNDEFINED/logs/fineract-platform.log```
```.response.json.swp```

if we need anyhow,
please try to add some documentation regarding those files

File: fineract-provider/src/main/java/org/apache/fineract/accounting/accrual/api/AccrualAccountingApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/accrual/api/AccrualAccountingApiResource.java]: I don't think, we need to add an extra blank line here

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java]: I don't think, we need to add an extra blank line here

File: fineract-provider/src/main/java/org/apache/fineract/accounting/financialactivityaccount/api/FinancialActivityAccountsApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/financialactivityaccount/api/FinancialActivityAccountsApiResource.java]: I don't think, we need to add an extra blank line here

File: fineract-provider/src/main/java/org/apache/fineract/organisation/workingdays/api/WorkingDaysApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/organisation/workingdays/api/WorkingDaysApiResource.java]: I am able to ```\n``` in swagger-ui,
try to remove all the unnecessary letters in the description

You can cross - verify for all the API's from swagger-UI

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java:50]: Indentation of ```@Tag``` annotation is not correct, please correct!

**Review by Sanyam96 — CHANGES_REQUESTED** [review]

**Comment by Sanyam96** [comment]
@kangbreder Please verify all the description of each and every API from swagger-UI.
and try to remove all the unnecessary letters/alphabets/characters from the code base

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java
Comment (kangbreder) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java:50]: For swagger's API documentation, here is the way to add description 
@Api(tags = {"Swagger Resource"})
@SwaggerDefinition(tags = {
    @Tag(name = "Swagger Resource", description = "Write description here")
})
public class ... {
}

That is what i got from swagger github issues.
does that mean i will have to unindent? I have not checked if that will work but right now, that description properly appears on the UI

Comment (Sanyam96) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java:50]: I'm seeing double ```tab-blank``` spaces in annotation 
```@Tag(name = "Floating Rates", description = "It lets you create, list, retrieve and upload the floating rates")```

File: fineract-provider/src/main/java/org/apache/fineract/accounting/accrual/api/AccrualAccountingApiResource.java
Comment (Sanyam96) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/accrual/api/AccrualAccountingApiResource.java]: Still seeing blank lines!

**Review by Sanyam96 — CHANGES_REQUESTED** [review]

**Comment by Sanyam96** [comment]
@kangbreder I am still seeing blank lines on API's 

Screenshot 2019-09-01 at 8 56 22 PM

You have also not given your explanation on https://github.com/apache/fineract/pull/629#issuecomment-524767338

**Comment by kangbreder** [comment]
> I don't think, we should require the below files in the codebase
> `catalina.base_IS_UNDEFINED/logs/fineract-platform.log`
> `.response.json.swp`
> 
> if we need anyhow,
> please try to add some documentation regarding those files

I think these files were auto generated during build some time back.

**Comment by vorburger** [comment]
kangbreder@ I just merged your two earlier smaller PRs. Will you rebase this one and resolve conflicts?

**Comment by Sanyam96** [comment]
@vorburger 
I think this PR is not completed yet and a lot of work is still needed to be done.

and this PR also needs to be tested before merging into the codebase.



@kangbreder can you please share the status of this PR?

**Comment by Sanyam96** [comment]
@kangbreder 
I can still able to view ```.response.json.swp``` file and ```catalina.base_IS_UNDEFINED/logs/fineract-platform.log``` file

Can you please remove these two files, as I think these are not relevant

**Review by Sanyam96 — CHANGES_REQUESTED** [review]
@kangbreder
I can still able to view .response.json.swp file and catalina.base_IS_UNDEFINED/logs/fineract-platform.log file

Can you please remove these two files, as I think these are not relevant

**Comment by kangbreder** [comment]
> kangbreder@ I just merged your two earlier smaller PRs. Will you rebase this one and resolve conflicts?

@vorburger I am currently resolving the conflicts. I will send an update soon. This PR is not ready for a merge yet. I am currently still working on this.

**Comment by vorburger** [comment]
@kangbreder @Sanyam96 how are you progressing with resolving the conflicts and making this ready to merge? If you were able to have complete Swagger documentation finished ASAP so that it could still make it into our 1.4.0 release that would be great!

**Comment by kangbreder** [comment]
I have been resolving the conflicts and they keep happening. I keep
resolving conflicts I had already resolved.

**Comment by vorburger** [comment]
> I have been resolving the conflicts and they keep happening. 
> I keep resolving conflicts I had already resolved.

Would you like assistance and have one of us help with resolving these conflicts (in new PR) ?  If so, perhaps this is something your GSoC mentor could support you with? I'm not sure who that was, was it @Sanyam96 or someone else? Otherwise, an active commiter such as e.g. @awasum or myself or @mohitsinha could perhaps take this over? I think it would be great if we could wrap up this PR by let's say end of next week latest.

**Comment by kangbreder** [comment]
Okay. I will connect with Sanyam this weekend.

**Comment by Sanyam96** [comment]
@kangbreder I would recommend completing the project or making a project to a stable state then resolving the conflicts at the final state.
so that we can release the done part of the docs in the next release

**Comment by kangbreder** [comment]
I have resolved the conflicts on the PR but having this build failure.
Here are the logs https://pastebin.com/DKzcGgZE
Please can i get some help with this?

**Comment by awasum** [comment]
Push the new changes to yours branch and they will show here. then someone might be able to help.

**Comment by kangbreder** [comment]
Okay. I have done that

**Comment by kangbreder** [comment]
I have resolved the conflicts on the PR but having this build failure.
Here are the logs https://pastebin.com/DKzcGgZE
Please can i get some help with this?

**Comment by vorburger** [comment]
> having this build failure. Please can i get some help with this?

This is because this PR, as is, also modifies *.gradle build script files.. and many more, check out the _Files change_ and _Commits_ tabs on this PR, and you'll see what I mean, it looks like this would revert unrelated previous changes - which I'm sure isn't what you are proposing here, but something in your local git operations has gone wrong. @Sanyam96 are you able to help @kangbreder to get this to a clean state? Please feel free to shout if you need me or @awasum to assist.

**Comment by vorburger** [comment]
@kangbreder the solution you are going for here seems to be to manually clean up the build script with additional commits? What you really want to do instead is to not have that many commits (155!), and make this PR only have your changes... have you tried to simply do a `git rebase` instead?

**Comment by kangbreder** [comment]
Yea i did a rebase before i got the build failure

**Comment by kangbreder** [comment]
> @kangbreder the solution you are going for here seems to be to manually clean up the build script with additional commits? What you really want to do instead is to not have that many commits (155!), and make this PR only have your changes... have you tried to simply do a `git rebase` instead?

@vorburger Yea i had done a rebase but after that, i got the build failure. So, i manually integrated the changes from develop branch.

**Comment by Sanyam96** [comment]
@kangbreder Can you please update the status of the project regarding the Jira ticket?
https://jira.apache.org/jira/browse/FINERACT-733

**Comment by Sanyam96** [comment]
@kangbreder I am also able to see lot of files, which are not related to your changes.
Can you please remove those and make this PR more relevant to the Swagger docs?

Also can you please squash your commits into a single commit?

File: .gitignore
Comment (Sanyam96) [inline .gitignore]: Please remove this

**Review by Sanyam96 — CHANGES_REQUESTED** [review]
@kangbreder Please remove all the files, which are not related to swagger
It seems that some commits from the master are merged into this PR, Maybe you can use
```git rebase``` command to remove all the commits, which are not associated with swagger-docs.

And kindly update the status of the swagger docs project.

**Comment by vorburger** [comment]
@kangbreder this PR is now looking much much better than ever before - congratulations!

There are just 2 conflicts, due to the recent merge of #678 - could you manually resolve those? Once those conflicts are resolved and when this passes the builds, then this looks good to me, and I'm happy to merge it ASAP - let's still get this into release 1.4.0! @Sanyam96 @awasum  does this have LGTM from you guys as well, as-is?

PS: One thing I would love to understand better, and I'm sure others would like to as well, is how this will actually be used? I'm sure to you this is "obvious", but to all those of us who have not been involved in the details this GSOC project, would it be possible to have some "doc" for this? It doesn't matter if initially it's given here or in [FINERACT-733](https://jira.apache.org/jira/browse/FINERACT-733) or may be somewhere on the Apache Confluence Wiki. Just to understand, once we merge this, what does this actually "do" - you know what I mean? I gather these two new Gradle plugins you are adding ([io.swagger.core.v3.swagger-gradle-plugin](https://github.com/swagger-api/swagger-core/tree/master/modules/swagger-gradle-plugin) and [org.hidetake.swagger.generator](https://github.com/int128/gradle-swagger-generator-plugin)) do... what - just generate the src/main/resources/swagger-ui/response.json file? Can I ask why we need 2 plugins to generate 1 file? :smiling_imp: Or do these to more

I can see that file already exists now, but I'm not clear where that's coming from, and how it was initially built.. does someone else know? @vishwasbabu or @Sanyam96 perhaps? Is that this PR contributes, in essence, really the automatic build of that file?

If that new src/main/resources/swagger-ui/response.json file is (now) a generated artifact, then why is it in Git, and generated under src/ instead of just being put in the right place under build/ so that it ends up in the WAR? Perhaps that could be contributed as a follow-up once we have merged this?

**Comment by kangbreder** [comment]
@vorburger Thank you. I have resolved the conflicts. 

The response.json file is the OpenAPI specification file which is used to generate the swagger-ui. 
The already existing response.json file in src/main/resources/swagger-ui/ is the spec generated with the request https://localhost:8443/fineract-provider/api/v1/swagger.json. Every time a modification is done on the swagger docs or a new swagger doc is added, this spec needs to be generated again at runtime and the response.json file as well updated so that the modifications can be seen on the UI. 

The io.swagger.core.v3.swagger-gradle-plugin is a plugin i had added to automatically generate the specs during build time and not at run time. This was so that every time one builds on the project, the response.json updates. But this did not work as expected. I had removed the plugin from the build script. I am wondering how its still in there. probably while i have rebasing and resolving conflicts, that came back in the script. I will remove the plugin. 

The other plugin org.hidetake.swagger.generator is used to generate code from the openAPI spec file using swagger-codegen. So this does not generate response.json. It rather uses response.json to generate client code. You

@vorburger i think you have a nice idea. Is it possible to get the file and the swagger-ui in the war? 
That is actually what is needed to be done. Presently the file is manually generated and updated each time.

**Comment by Sanyam96** [comment]
@kangbreder thank you for making the PR more stable.

I'll test your changes very soon so that we can merge and release your changes in the next cycle release cycle.

@vorburger 
I also agree with you regarding usability.
I think we should document somewhere, maybe at a centric place and also provide the link in the readme so that other devs can refer the swagger-docs and how to use swagger docs in the future.


I am also not sure, about the plugins, do we require it or not, right now.
I will review this PR, and provide my feedback and required changes to this PR.

**Comment by vorburger** [comment]
> @vorburger Thank you. I have resolved the conflicts.

following the merge of #682, there now is a new conflict again already... will you resolve, again?

I think to avoid that happening again and again, we should try to get a minimal version of this PR merged ASAP, and then incrementally iterate on improving it in separate follow-up PRs.

> The io.swagger.core.v3.swagger-gradle-plugin is a plugin i had added to automatically generate the specs during build time and not at run time. This was so that every time one builds on the project, the response.json updates. But this did not work as expected. I had removed the plugin from the build script. I am wondering how its still in there. probably while i have rebasing and resolving conflicts, that came back in the script. I will remove the plugin.

OK, cool. From my side, we could perhaps merge this PR after you've done that (but @Sanyam96 more than me should fully review and LGTM this).

> @vorburger i think you have a nice idea. Is it possible to get the file and the swagger-ui in the war?

Yeah, I think it should be. @Sanyam96 can probably work with you towards that. Again, I would suggest that we get this PR to a state where it can be merged even if it can still be improved in follow-up you can raise in separate new PRs.

**Comment by awasum** [comment]
@Sanyam96 , @vorburger , Seems conflicts are now fixed by @kangbreder  and Travis is passing. Merge this when yo feel comfortable... instead create more issues and enhancement tickets related to this feature on Jira for @kangbreder  to pick up and fix gradually.

**Review by vorburger — APPROVED** [review]
@Sanyam96 does this have LGTM from you? From my side, I'm fully aligned with @awasum - let us finally get this merged ASAP, and then gradually improve upon it (I do understand the work expected here is not fully done yet).

I'm going to try to have this one automagically rebased using #691, curious if this will work (1st time):

/rebase

File: fineract-provider/build.gradle
Comment (vorburger) [inline fineract-provider/build.gradle:63]: @kangbreder do you happen to remember why you had to add `apply plugin: 'java'` here? Does it work without it as well? I guess it's a bit strange to have a project use BOTH `apply plugin: 'java-library'` AND `apply plugin: 'java'` - it's either or the other.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java:50]: @Sanyam96 I would not worry too much about this kind of thing (whitespace and indent), we're about to have a general solution for this, please watch [FINERACT-821](https://issues.apache.org/jira/browse/FINERACT-821). (Help from you on that one would be very welcome BTW!)

**Comment by vorburger** [comment]
FYI: /rebase does not actually seem to work just yet, see [FINERACT-829](https://issues.apache.org/jira/browse/FINERACT-829) / [INFRA-19719](https://issues.apache.org/jira/browse/INFRA-19719). We can manually just manually rebase it, once we have the LGTM from @Sanyam96 on this one. (Or we just could merge it with a merge commit; not my favourite, but also fine.)

File: fineract-provider/src/main/java/org/apache/fineract/mix/api/MixTaxonomyMappingApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/mix/api/MixTaxonomyMappingApiResource.java:35]: remove this import
I think it should be covered in line 36

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/search/api/SearchApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/search/api/SearchApiResource.java:98]: Please add space between
```suggestion
    public String searchData(@Context final UriInfo uriInfo, @QueryParam("query")  @ApiParam(value = "query") final String query,
```

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/self/shareaccounts/api/SelfShareAccountsApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/self/shareaccounts/api/SelfShareAccountsApiResource.java:158]: Why these annotations is commented??

File: fineract-provider/build.gradle
Comment (Sanyam96) [reply fineract-provider/build.gradle:63]: I also think this is not relevant to this swagger changes
@kangbreder can you please verify this.

File: fineract-provider/build.gradle
Comment (Sanyam96) [inline fineract-provider/build.gradle:45]: I think this plugin would not help to generate the swagger.json automatically during project build
please remove this.

File: fineract-provider/build.gradle
Comment (Sanyam96) [inline fineract-provider/build.gradle:552]: @kangbreder  Can you please clarify about these changes?
I am not sure how this code will help us?.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/api/ReportMailingJobRunHistoryApiResource.java
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/api/ReportMailingJobRunHistoryApiResource.java:46]: Why it is removed?

**Review by Sanyam96 — CHANGES_REQUESTED** [review]

**Comment by Sanyam96** [comment]
@kangbreder Please do the following changes in the PR.

1. Please remove all the changes from the build.gradle file.
2. Please make the changes mentioned above in my review.
3. Please squash all the commits into a single commit.

@vorburger @awasum 
I think apart from the above-mentioned changes Everything else Looks Good To Me.
After Kang made all the changes, we can work to merge this PR.

File: fineract-provider/build.gradle
Comment (kangbreder) [reply fineract-provider/build.gradle:63]: > @kangbreder do you happen to remember why you had to add `apply plugin: 'java'` here? Does it work without it as well? I guess it's a bit strange to have a project use BOTH `apply plugin: 'java-library'` AND `apply plugin: 'java'` - it's either or the other.

@vorburger I have not checking if it works without the plugin. My Ubuntu 18 has failed to boot and i have been trying to correct the problem. Because of that, I have not been able to check that yet

File: fineract-provider/build.gradle
Comment (kangbreder) [reply fineract-provider/build.gradle:552]: > @kangbreder Can you please clarify about these changes?
> I am not sure how this code will help us?.

@Sanyam96  have you tried using this to see how it works? Are you sure you really want me to remove all changes made on the build.script. Are you sure about that?

Comment (kangbreder) [reply fineract-provider/build.gradle:552]: > @kangbreder Can you please clarify about these changes?
> I am not sure how this code will help us?.

@Sanyam96 I think i had explained the usage of this in my gist https://gist.github.com/kangbreder/034f47e2e8015cee10b48b7c5f1b8df1, generating client libraries and language-specific SDKs in languages such as java and angular from swagger specification file. This was part of the goals of the project listed in https://mifosforge.jira.com/wiki/spaces/RES/pages/812810251/Google+Summer+of+Code+2019+Ideas. 

I clarified the usage of this this to @vorburger the other day. I had removed the other plugin which was added for generation of spec file during build. Since that did not work, it has been removed it from the PR. 

I suggest you verify that again.
 My ubuntu 18 will not boot.
 I have been trying to solve the problem. Because of that, i have not been able to do the modifications.

Comment (Sanyam96) [reply fineract-provider/build.gradle:552]: Thanks for the clarification @kangbreder .

Please do other requested changes other than build.gradle.

**Comment by vorburger** [comment]
@kangbreder because I'd like to help you avoid running into painful whitespace related merge issues due to #688 which I'm currently working on and would like to wrap up today, here's what I'm going to do about this PR:

I'll local pull this, and make the changes which @Sanyam96 suggests, and merge it ASAP. (This will allow me to then rebase #688 and do the overall whitespace clean up, which will unblock #689 which I'd like to get in rather sooner than later.)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/self/shareaccounts/api/SelfShareAccountsApiResource.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/self/shareaccounts/api/SelfShareAccountsApiResource.java:158]: @Sanyam96 FYI I've fixed this one in my swagger-squash branch which I'll merge ASAP as much as possible. @kangbreder the problem is that SelfShareAccountsApiResourceSwagger.GetShareAccountResponse seems to be missing- would you like to raise a small new PR just adding that and fixing this?

**Comment by vorburger** [comment]
Closing this PR as having been replaced with #695 (which is basically this, with Authorship by Kang retained in the git history of course), just squashed all into 1 single commit, followed up by a commit from me which implements the requested changes.

[FINERACT-835](https://jira.apache.org/jira/browse/FINERACT-835) now includes a number of more specific new sub-tasks with proposed follow-ups.

**Pull request closed without merging**
